crates/federate/src/util.rs

```diff
@@ -0,0 +1,177 @@
+use reqwest::Url;
+use serde_json::Value;
+use std::{future::Future, pin::Pin, sync::Arc, time::Duration};
+use tokio::{task::JoinHandle, time::sleep};
+use tokio_util::sync::CancellationToken;
+
+pub struct CancellableTask<R: Send + 'static> {
+  f: Pin<Box<dyn Future<Output = Result<R, anyhow::Error>> + Send + 'static>>,
+}
+
+impl<R: Send + 'static> CancellableTask<R> {
+  /// spawn a task but with graceful shutdown
+  pub fn spawn<F>(
+    timeout: Duration,
+    task: impl FnOnce(CancellationToken) -> F,
+  ) -> CancellableTask<R>
+  where
+    F: Future<Output = Result<R>> + Send + 'static,
+  {
+    let stop = CancellationToken::new();
+    let task = task(stop.clone());
+    let task: JoinHandle<Result<R>> = tokio::spawn(async move {
+      match task.await {
+        Ok(o) => Ok(o),
+        Err(e) => {
+          tracing::error!("worker errored out: {e}");
+          // todo: if this error happens, requeue worker creation in main
+          Err(e)
+        }
+      }
+    });
+    let abort = task.abort_handle();
+    CancellableTask {
+      f: Box::pin(async move {
+        stop.cancel();
+        tokio::select! {
+            r = task => {
+                Ok(r.context("could not join")??)
+            },
+            _ = sleep(timeout) => {
+                abort.abort();
+                tracing::warn!("Graceful shutdown timed out, aborting task");
+                Err(anyhow!("task aborted due to timeout"))
+            }
+        }
+      }),
+    }
+  }
+
+  /// cancel the cancel signal, wait for timeout for the task to stop gracefully, otherwise abort it
+  pub async fn cancel(self) -> Result<R, anyhow::Error> {
+    self.f.await
+  }
+}
+
+/// assuming apub priv key and ids are immutable, then we don't need to have TTL
+/// TODO: capacity should be configurable maybe based on memory use
+pub(crate) async fn get_actor_cached(
+  pool: &mut DbPool<'_>,
+  actor_type: ActorType,
+  actor_apub_id: &Url,
+) -> Result<Arc<SiteOrCommunityOrUser>> {
+  static CACHE: Lazy<Cache<Url, Arc<SiteOrCommunityOrUser>>> =
+    Lazy::new(|| Cache::builder().max_capacity(10000).build());
+  CACHE
+    .try_get_with(actor_apub_id.clone(), async {
+      let url = actor_apub_id.clone().into();
+      let person = match actor_type {
+        ActorType::Site => SiteOrCommunityOrUser::Site(
+          Site::read_from_apub_id(pool, &url)
+            .await?
+            .context("apub site not found")?
+            .into(),
+        ),
+        ActorType::Community => SiteOrCommunityOrUser::UserOrCommunity(UserOrCommunity::Community(
+          Community::read_from_apub_id(pool, &url)
+            .await?
+            .context("apub community not found")?
+            .into(),
+        )),
+        ActorType::Person => SiteOrCommunityOrUser::UserOrCommunity(UserOrCommunity::User(
+          Person::read_from_apub_id(pool, &url)
+            .await?
+            .context("apub person not found")?
+            .into(),
+        )),
```

Comment:
I dont think the `SiteOrCommunityOrUser` enum is necessary, because all variants implement ApubActor trait which has the method `read_from_apub_id`. So you only need to ensure that cached values implement this trait.

Reply:
I'm storing and returning mixed values of this struct in get_actor_cached(). I can probably change that to dynamic invokation by using `Box<dyn Actor>`, let me see.

Reply:
Unfortunately, that doesn't seem to work. In order to create a `Box<dyn Actor>`, I would need to specify the associated types `Box<dyn Actor<DataType=LemmyContext, Kind=?, Error=LemmyError>>`. But the `Kind` type differs for each different Actor, so it's impossible to create that box for different types of actors.

Reply:
I could replace SiteOrCommunityOrUser with a thinner struct that just wraps Box<dyn Actor>, and sets `Kind` to `()`. That seems a bit hacky though.